### PR TITLE
feat: save peers as enr

### LINKED ISSUE #1384: Save peers as enr isntead of NodeRecord
### Describe the feature

On drop we're saving a file with all `NodeRecords`:

https://github.com/paradigmxyz/reth/blob/c415221876110914699366a85bfd4fb2e72a62b8/bin/reth/src/node/mod.rs#L391-L394

https://github.com/paradigmxyz/reth/blob/c415221876110914699366a85bfd4fb2e72a62b8/crates/net/network/src/manager.rs#L299-L301

however, some peers announce a fork id in discovery which we should also save

https://github.com/paradigmxyz/reth/blob/c415221876110914699366a85bfd4fb2e72a62b8/crates/net/network/src/peers/manager.rs#L755-L756

## TODO

* convert `Peer -> Enr` with "eth" field, ref https://github.com/paradigmxyz/reth/blob/c415221876110914699366a85bfd4fb2e72a62b8/crates/net/discv4/src/lib.rs#L443-L455

* ideally, move this conversion to new function in `NodeRecord` `fn int_enr_builder(self) -> EnrBuilder`
* update file reading code in `NetworkConfig`

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by leruaa**
@mattsse I started to work on the issue #1384, but I fail to find where file reading code is in NetworkConfig. Can you help?

Thanks

Closes #1384

### CODE REVIEW

File: bin/reth/src/node/mod.rs [hunk 1]
```diff
@@ -376,6 +377,7 @@ impl Command {
 async fn run_network_until_shutdown<C>(
     shutdown: reth_tasks::shutdown::Shutdown,
     network: NetworkManager<C>,
+    secret_key: &SecretKey,
```
Comment (mattsse) [inline bin/reth/src/node/mod.rs:380]: the network already has this so we don't actually need it if we let the network do it.

**Review by mattsse — COMMENTED** [review]
I didn't realize that we still need to sign the enr because we're operating with NodeRecord and Enrbuilder does not have a way to create an Enr with the Signature directly so this feature isn't actually doable atm.
We'd need to find a way to create an `Enr` with the signature directly first...

I'm so sorry for wasting your time on this...

Blocked because `Enr` can't be created without signing it...

this could be unblocked by adding `build_with_public_key` here.

https://github.com/sigp/enr/blob/master/src/builder.rs#L144

imo this is reasonable because there's already a `add_public_key` function https://github.com/sigp/enr/blob/master/src/builder.rs#L136

**Comment by leruaa** [comment]
@mattsse Do you want me to create the issue to add `build_with_public_key()`?

**Comment by leruaa** [comment]
So if I understood correctly, we need to manually impl `RlpEncode` (and `RlpDecode`) on `NodeRecord` in order to match `Enr` implementation, and be able to build an `Enr` from the resulting rlp-encoded base64.

@mattsse Can you confirm If I understood correctly what's need to be done?

**Comment by mattsse** [comment]
yeh, we basically want a `NodeRecord::to_base64` that behaves like `Enr::to_base64` which is base64(rlp(enr))

so we need a custom rlp encode for NodeRecord that mirrors:

```rust
    fn rlp_append(&self, s: &mut RlpStream) {
        s.begin_list(self.content.len() * 2 + 2);
        s.append(&self.signature);
        s.append(&self.seq);
        // must use rlp_content to preserve ordering.
        for (k, v) in &self.content {
            // Keys are byte data
            s.append(k);
            // Values are raw RLP encoded data
            s.append_raw(v, 1);
        }
    }
```
and also accepts additional fields

**Comment by gakonst** [comment]
just caught up on the Sigma Prime thread and +1 on Matt's suggestion, thanks for running this down @leruaa

**Comment by gakonst** [comment]
Did this end up making it into the ENR crate @leruaa ? If not, should we close this?

**Comment by gakonst** [comment]
closing per discussion in the issue

**Pull request closed without merging**